File: Dynamic_Programming/931.Minimum_Falling_Path_Sum/C++/931.Minimum_Falling_Path_Sum.cpp
class Solution {
public:
    int minFallingPathSum(vector<vector<int>> & matrix) 
    {
        //==============================================//
        // Input type:                                  //
        //   - vector<vector<int>> ref. (matrix vector) //
        // Return type:                                 //
        //   - int (result minimun summary value)       //
        //==============================================//

        /*Initialize*/
        ///// Length of rows, columns /////
        int len_rows = matrix.size(), len_cols = (matrix[0]).size();

        ///// Record dynamic-programming (i.e. DP) vector /////
        vector<vector<int>> record_dp(matrix.begin(), matrix.end());

        ///// Result minimun summary value /////
        int res_min_sum = INT_MAX;


        /*Dynamic-programming based loop traversal*/
        for (int record_rows = (len_rows - 2); (record_rows >= 0); (record_rows--)) //Rows
        {
            for (int record_cols = 0; (record_cols < len_cols); (record_cols++)) //Columns
            {
                int next_min_val = (record_dp[(record_rows + 1)])[record_cols]; //Next minimun value

                ///// Check if the next columns index exceed boundary or not /////
                if ((record_cols - 1) >= 0)
                {
                    next_min_val = min((record_dp[(record_rows + 1)])[(record_cols - 1)], next_min_val); //Keep updating, overwriting
                }
                else { ; }

                ///// Check if the next columns index exceed boundary or not /////
                if ((record_cols + 1) < len_cols)
                {
                    next_min_val = min((record_dp[(record_rows + 1)])[(record_cols + 1)], next_min_val); //Keep updating, overwriting
                }
                else { ; }

                (record_dp[record_rows])[record_cols] += next_min_val; //Keep updating, accumulating

                ///// Check if the current row index matched conditions or not /////
                if (! record_rows)
                {
                    res_min_sum = ((record_dp[record_rows])[record_cols] < res_min_sum) ? 
                                   (record_dp[record_rows])[record_cols] : res_min_sum; //Keep updating, overwriting
                }
                else { ; }

            } //Columns

        } //Rows

        return (res_min_sum == INT_MAX) ? (matrix[0])[0] : res_min_sum;
    }
};

File: Dynamic_Programming/931.Minimum_Falling_Path_Sum/Python/931.Minimum_Falling_Path_Sum.py
class Solution(object):
    def minFallingPathSum(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: int
        """
        #======================================================#
        # Dynamic-programming (i.e. DP) based traversal method #
        #======================================================#

        ############
        #Initialize
        ##### Length of rows, columns #####
        len_rows, len_cols = len(matrix), len(matrix[0])

        ##### Record dynamic-programming (i.e. DP) array #####
        record_dp = deepcopy(matrix)

        ##### Result minimun summary value #####
        res_min_sum = float("inf")


        ####################################################
        #Dynamic-programming (i.e. DP) based loop traversal
        for rows_idx in range((len_rows - 2), (-1), (-1)):
            for cols_idx in range(len_cols):
                next_min_val = (record_dp[(rows_idx + 1)])[cols_idx] #Next minimun value

                ##### Check if the next columns index exceed boundary or not #####
                if ((cols_idx - 1) >= 0):
                    next_min_val = min((record_dp[(rows_idx + 1)])[(cols_idx - 1)], next_min_val) #Keep updating/overwriting
                else:
                    pass

                ##### Check if the next columns index exceed boundary or not #####
                if ((cols_idx + 1) < len_cols):
                    next_min_val = min((record_dp[(rows_idx + 1)])[(cols_idx + 1)], next_min_val) #Keep updating/overwriting
                else:
                    pass

                (record_dp[rows_idx])[cols_idx] += next_min_val #Keep updating/accumulating

                ##### Check if the current rows index matched conditions or not #####
                if (not rows_idx):
                    res_min_sum = min((record_dp[rows_idx])[cols_idx], res_min_sum) #Keep updating/overwriting
                else:
                    pass

        return ((record_dp[0])[0] if (res_min_sum == float("inf")) else res_min_sum)
